### ISSUE TITLE
Crash on device after VS/Xamarin.iOS update

### ISSUE THREAD

**Issue by dpjustdoit** [xamarin/xamarin-macios#6750]
### Steps to Reproduce

After VS/Xamarin.iOS update I got constant crashes on devices (iOS12.3, iOS12.4, may be others) in different `UIViewController` subclasses. On simulator everything still works as expected. VS that worked was 7.8.3. I tried different 8.x.x versions and the latest one.

Xamarin's `UIViewController` subclasses are collected (disposed) for some reason.

### Expected Behavior

It should work on device as well as on simulator.

### Actual Behavior

Crashes

### Environment

```
=== Visual Studio Community 2019 for Mac ===

Version 8.2.2 (build 21)
Installation UUID: 4c33eaa4-0aca-438d-b784-d5388862cf78
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.14.0.85 (d16-2-fix-mac-updates / 86973979)

	Package version: 600000319

=== Mono Framework MDK ===

Runtime:
	Mono 6.0.0.319 (2019-02/52203786470) (64-bit)
	Package version: 600000319

=== NuGet ===

Version: 5.1.0.6013

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Version: 2.1.12
SDK: /usr/local/share/dotnet/sdk/2.1.701/Sdks
SDK Version: 2.1.701
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/msbuild/Current/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.11
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 10.3 (14492.2)
Build 10G8

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin.Android ===

Version: 9.4.1.0 (Visual Studio Community)
Commit: xamarin-android/d16-2/ebd65c9
Android SDK: /Users/dp/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.2
SDK Build Tools Version: 28.0.3

Build Information: 
Mono: mono/mono/2019-02@e6f5369c2d2
Java.Interop: xamarin/java.interop/d16-2@d64ada5
LibZipSharp: grendello/LibZipSharp/d16-2@caa0c74
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-2@6f6c969

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/dp/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android Device Manager ===

Version: 1.2.0.58
Hash: d03e3ea
Branch: remotes/origin/d16-2~1
Build date: 2019-07-30 21:29:36 UTC

=== Xamarin.iOS ===

Version: 12.14.0.114 (Visual Studio Community)
Hash: c6691164
Branch: d16-2
Build date: 2019-07-29 15:54:05-0400

=== Xamarin Designer ===

Version: 16.2.0.379
Hash: 3f0652957
Branch: remotes/origin/d16-2
Build date: 2019-08-01 17:32:50 UTC

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 802020021
Git revision: d2ffd09000c5d24a206b715b63d409e8e965689a
Build date: 2019-08-02 00:25:49+00
Build branch: release-8.2
Xamarin extensions: 4cc9cbbe68baaeb6bc654f36e057b1bd124f3f65

=== Operating System ===

Mac OS X 10.14.6
Darwin 18.7.0 Darwin Kernel Version 18.7.0
    Thu Jun 20 18:42:21 PDT 2019
    root:xnu-4903.270.47~4/RELEASE_X86_64 x86_64
```

### Build Logs

Logs are similar:
```
ObjCRuntime.RuntimeException: Failed to marshal the Objective-C object 0x11bbe8a00 (type: __ABT_XViewController). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance (because the type 'YYY.iOS.Support.XViewController' does not have a constructor that takes one IntPtr argument).
```

PLEASE ADVISE
P.S. Can it be caused by some project settings ?

**Comment by rolfbjarne** [comment]
Thank you for your feedback!

For us to investigate this further, could you please provide a test case we can use to reproduce the problem?

We look forward to hearing from you!

**Comment by dpjustdoit** [comment]
@rolfbjarne 
Thanks for the answer. The thing with this issue is that it seems to be local. I'm not able to reproduce it on the sample project and I can't send you our project.
Here is the issue I faced and not able to fix, MAY BE you have some ideas what else I can try…

After VS/Xamarin update I got constant crashes on device (iOS12.3, iOS12.4, may be others) in different `UIViewController` subclasses (Important: no code changes were made). On simulator everything still works as expected. VS that worked was 7.8.3.

Logs are similar, but in different classes:
```
ObjCRuntime.RuntimeException: Failed to marshal the Objective-C object 0x11bbe8a00 (type: __ABT_XViewController). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance (because the type 'YYY.iOS.Support.XViewController' does not have a constructor that takes one IntPtr argument).
```
Xamarin’s `UIViewController` subclasses are collected (disposed) for some reason, while native controllers are still alive. Adding constructor with IntPtr doesn’t work as there are properties linked to native ones, so Handles will be lost (Buttons, Labels, etc.)
I tried installation of different VS 8.x.x versions and the latest one.
Tried removing and reinstalling different Frameworks versions (Mono.framework, Xamarin.iOS.framework, /usr/local/share/dotnet/dotnet/).
Tried cloning the repo.
Tried simple sample project and don’t see this wrong `UIViewController` disposal (we can suggest that it is caused by some project settings, but there were no changes and I have no idea which setting can cause such behavior).
Also updated (checked) and after that reset iOS on my iPhone to make sure the issue isn’t there.

Several months ago, we faced similar issue and the solution was to downgrade back to 7.8.3. But 7.8.3 doesn’t support Xcode Version 10.3 (10G8) and iOS12.4. Also I am not able to downgrade my iPhone - https://www.macrumors.com/2019/07/31/apple-stops-signing-ios-12-3-1/

To summarize the issue:
It is only on device. The problem is not in particular `UIViewController` subclass, the problem is that C# `UIViewController`s are collected (disposed) almost instantly, while native controllers are still alive.

Do you have any assumptions or ideas?
Best regards

**Comment by rolfbjarne** [comment]
I have a diagnostic idea: try to force the GC to execute frequently, add something like this to your Main method:

```csharp
var thread = new Thread (() =>
{
    while (true) {
        Thread.Sleep (5000);
        GC.Collect ();
    }
});
thread.Start ();
```

in particular it would be interesting to see if this makes the problem occur in the simulator as well.

**Comment by dpjustdoit** [comment]
With this forced collection on simulator works fine.
Actually, disposed managed Controllers are in the UIView hierarchy.

**Comment by rolfbjarne** [comment]
Another idea to try to make it fail in the simulator: add `--registrar:static` to the additional mtouch arguments in the project's iOS Build options. Also make the Linker behavior option identical between your device and simulator builds.

Both of these should make the simulator and device builds more similar.

**Comment by dpjustdoit** [comment]
It worked and now the simulator has the same crashes.

**Comment by dpjustdoit** [comment]
I'll debug deeper, but I observe even managed `UIViewController` set to be `UIWindow`'s RootViewController is disposed. Is there any next step on your mind?

**Comment by rolfbjarne** [comment]
No, something is quite wrong if UIWindow's RootViewController is being disposed, that really shouldn't happen. This also means I have no idea what could be happening, so without a test project I can use to reproduce I don't really have any suggestions about what to do. Maybe try commenting out code from your real project to try to figure out the significant difference to a test project that makes things break?

**Comment by dpjustdoit** [comment]
Thanks for your response, Rolf.
What we have for now: this issue is reproduced for our project on different devices and simulator (with `--registrar:static` and Linker behavior changed). Everything works as expected on simple sample projects. To be sure we installed latest macOS/Xcode/VS on the 'clean' Mac, and the issue is present in such case. So looks like some project setting, dependency or may be code (it will be very strange) causes wrong managed `UIViewController` disposing behavior.
We have quite big project, so it will be quite difficult to find the cause by just guessing (changing settings, dependencies, code).
We need to discuss the possibility somehow send you project or part of it and our further steps.

**Comment by chamons** [comment]
We have not received the requested information. Wi

If you are still experiencing this issue please provide all the requested information then click the Reopen Issue button.

Thanks!

**Comment by ShamineDsouza** [comment]
Hi,

 I am facing the same issue , after upgrading my Device IOS to 12.4.8. Please advice 
I am currently doing a xamrin cross platform application. It was working fine till yesterday .     Should i set some version while deploying

**Comment by chamons** [comment]
@ShamineDsouza - Please consider opening a new issue with detailing including the build log, crash report (if any) and preferably a sample project.
